File: frontend/src/app/models/transaction.ts
import { Category } from './category';
import { Account } from './account';
import { Rule } from './rule';

export enum TransactionType {
    Transfer = 0,
    Expense,
    Income,
    Correction
}

export interface Transaction {
    id?: number | null;
    opdate: string;
    account?: Account | null;
    account_balance?: number;
    credit: number;
    recipient?: Account | null;
    recipient_balance?: number;
    debit: number;
    category?: Category | null;
    currency: string;
    party?: string;
    details: string;
    type: TransactionType;
    bg?: string;
}

export interface TransactionImport extends Transaction {
    selected: boolean;
    rule: Rule;
}

File: backend/app/app/schemas/transaction.py
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel

from .category import Category
from .account import Account

class TransactionBase(BaseModel):
    opdate: datetime
    category: Category = None
    account: Account = None
    credit: Decimal
    recipient: Account = None
    debit: Decimal
    currency: str = None
    party: str = None
    details: str = None

class TransactionCreate(TransactionBase):
    pass

class TransactionUpdate(TransactionBase):
    id: int

class TransactionImport(TransactionBase):
    id: int = None
    type: int

class Transaction(TransactionBase):
    id: int
    type: int
    account_balance: Decimal = None
    recipient_balance: Decimal = None
    bg: str

    class Config:
        orm_mode = True
